### PR TITLE
Manage tag hooks from gogs

### PR DESCRIPTION
**Pull request by bsauvajon**
Clean re-submit of the PR #1448

### CODE REVIEW

**Comment by morph027** [comment]
Just deployed it in my lab (cherry picking ;) ) and it works like charm. Thank you!

**Comment by bradrydzewski** [comment]
@bsauvajon sorry I haven't merged this sooner. I completely lost track of this pull request :frowning:  ... if you don't mind resolving conflicts I can merge right away. Thanks!

**Comment by bsauvajon** [comment]
Hi,
Conflict is resolved, but now the PR shows every commit since February. Can you merge or do you have to make a new clean PR ?

**Comment by tboerger** [comment]
Now you messed up the pr :)
You should always rebase against master.

**Comment by bsauvajon** [comment]
Ok, that's what I missed, looks much better
Thanks

File: model/const.go [hunk 1]
```diff
@@ -5,6 +5,7 @@ const (
 	EventPull   = "pull_request"
 	EventTag    = "tag"
 	EventDeploy = "deployment"
+	EventBranch = "branch"
```
Comment (bradrydzewski) [inline model/const.go]: I think I overlooked one thing. Introducing a new Event type is actually quite a bit of work and is used in a lot of different places, including for security purposes to determine when secrets are injected.

Can you provide more information about the branch event type? Could we use the push event type for Gogs instead? Just want to understand if we __need__ a new event type

**Comment by bradrydzewski** [comment]
I'm good to merge if we remove the `EventBranch` and replace with `EventPush`.  The events are defined not by the hooks but by the build process. The build process does not differentiate between commits and branches and are handled the same. Introducing a new event type is not required, and would be a massive change since event types have security implications as well

**Comment by bsauvajon** [comment]
This event is not really required, I thought it was just a cleaner way to manage branch events sent by gogs. If I used EventPush, as the workload sent by gogs does not contain all the necessary informations for a push, no build should happen but there will be an error in the logs.

**Comment by bradrydzewski** [comment]
We can use the branch hook but should not create a new `EventBranch` enumeration. These enumerations are for build events, not hook events, and adding a new event is not an option so we'll need to adjust as follows:

```diff
 	EventPull   = "pull_request"
 	EventTag    = "tag"
 	EventDeploy = "deployment"
-	EventBranch = "branch"
```

and then

```
	switch {
	case hook.RefType == "tag":
		eventType = model.EventTag
		message = "Tag " + hook.Ref
	case hook.RefType == "branch":
-		eventType = model.EventBranch
+		eventType = model.EventPush
	default:
		eventType = model.EventPush
		message = hook.Commits[0].Message
	}
```

**Comment by bsauvajon** [comment]
I've made some tests using EventPush when a branch is created, as you suggest. But it triggers a build which fails as the workload does not contain the required informations.
If I use EventTag instead, a build is triggered too, and it is successful, but I don't know if it is a good idea to have a build each time a branch is created ?

File: remote/gogs/helper.go [hunk 2]
```diff
@@ -62,13 +62,28 @@ func buildFromPush(hook *pushHook) *model.Build {
...
+	var eventType string
+	var message string
+
+	switch {
+	case hook.RefType == "tag":
+		eventType = model.EventTag
+		message = "Tag " + hook.Ref
+	case hook.RefType == "branch":
+		eventType = ""
+	default:
```
Comment (bradrydzewski) [inline remote/gogs/helper.go]: setting `eventType = ""` seems a bit undesirable because I'm not sure what the side effect will be. So more from a readability / understandability issue I'd like to avoid this.

Instead I'd recommend we do something like this:

```
func (c *client) Hook(r *http.Request) (*model.Repo, *model.Build, error) {
	var (
		err   error
		repo  *model.Repo
		build *model.Build
	)

	switch r.Header.Get("X-Gogs-Event") {
	case "push":
		var push *pushHook
		push, err = parsePush(r.Body)
		if err == nil && push.RefType != "branch" {
			repo = repoFromPush(push)
			build = buildFromPush(push)
		}
	}
	return repo, build, err
}
```

**Comment by bkleef** [comment]
@bradrydzewski / @bsauvajon any update on this? This makes me sad:

```
$ /usr/bin/docker tag quay.io/profects/cfssl:latest quay.io/profects/cfssl:$$TAG
Error parsing reference: "quay.io/profects/cfssl:$$TAG" is not a valid repository/tag
[info] build failed (exit code 1)
```

**Comment by bradrydzewski** [comment]
@bkleef there are still unresolved comments from my codereview. This is be merged once resolved

**Comment by bsauvajon** [comment]
Sorry, didn't saw the notification about your last comment, I'll work on it asap

**Comment by bkleef** [comment]
@bsauvajon thanks a lot for your update! @bradrydzewski any chance to review changes?

**Comment by cusspvz** [comment]
mentioning the issue #1412 just to bind them.

@bsauvajon thanks for diving on it!
@bradrydzewski can you please review?

**Comment by bkleef** [comment]
@bradrydzewski looks like @tboerger approved the changes of @bsauvajon, can we please close this PR and enjoy them in Drone master branch? Thanks a million!

File: remote/gogs/helper.go [hunk 3]
```diff
@@ -70,13 +70,26 @@ func buildFromPush(hook *pushHook) *model.Build {
...
+		message = hook.Commits[0].Message
+	}
+
 	return &model.Build{
-		Event:     model.EventPush,
+		Event:     eventType,
 		Commit:    hook.After,
 		Ref:       hook.Ref,
 		Link:      hook.Compare,
 		Branch:    strings.TrimPrefix(hook.Ref, "refs/heads/"),
```
Comment (bradrydzewski) [inline remote/gogs/helper.go]: we should change this to `"refs/tags/%s"`  when the hook type is tag. We can do this in a separate pull request, since I don't want to delay this merge any further :)

Comment (appleboy) [inline remote/gogs/helper.go]: ref: https://github.com/drone/drone/pull/1800
